### PR TITLE
Add relocation for deviceIndex builtin constant

### PR DESCRIPTION
**Pull request by csyonghe**
Based on #457, this PR makes deviceIndex  a relocatable constant.

### CODE REVIEW

**Comment by amdvlk-admin** [comment]
Can one of the admins verify this patch?

**Comment by JacobHeAMD** [comment]
Ok to test

**Comment by JacobHeAMD** [comment]
test this please

**Comment by csyonghe** [comment]
@JacobHeAMD  This PR requires a llvm change to build. https://reviews.llvm.org/D76440
Once that is in, the tests should pass.

**Comment by JaxLinAMD** [comment]
retest this please

File: llpc/util/llpcElfWriter.cpp [hunk 1]
```diff
@@ -1267,6 +1267,28 @@ void FixUpDescriptorOffsetRelocation(
     WriteRelocationConstant(pWriter, relocEntry.offset, offset);
 }
 
+// =====================================================================================================================
+// Fixup a device index relocation ($devIdx symbol) in the ELF with an actual device index value.
+template<class Elf>
+void FixupDeviceIndexRelocation(
+    ElfWriter<Elf> *pWriter,         // [InOut] Writer to target ELF
+    Context *pContext,               // Pipeline compilation context
```
Comment (amdrexu) [inline llpc/util/llpcElfWriter.cpp]: Add a tag "// [in] ..."

File: llpc/util/llpcElfWriter.cpp [hunk 2]
```diff
@@ -1267,6 +1267,28 @@ void FixUpDescriptorOffsetRelocation(
     WriteRelocationConstant(pWriter, relocEntry.offset, offset);
 }
 
+// =====================================================================================================================
+// Fixup a device index relocation ($devIdx symbol) in the ELF with an actual device index value.
+template<class Elf>
+void FixupDeviceIndexRelocation(
+    ElfWriter<Elf> *pWriter,         // [InOut] Writer to target ELF
```
Comment (amdrexu) [inline llpc/util/llpcElfWriter.cpp]: [InOut] -> [in,out]

File: llpc/patch/llpcPatchInOutImportExport.cpp [hunk 3]
```diff
@@ -2352,7 +2352,14 @@ Value* PatchInOutImportExport::PatchVsBuiltInInputImport(
     case BuiltInSubgroupLocalInvocationId:
         return GetSubgroupLocalInvocationId(pInsertPos);
     case BuiltInDeviceIndex:
-        return ConstantInt::get(Type::getInt32Ty(*m_pContext), m_pPipelineState->GetDeviceIndex());
+        if (m_pPipelineState->GetBuilderContext()->BuildingRelocatableElf())
+        {
+            return EmitRelocationConstant(m_pContext, pInsertPos, "$devIdx");
```
Comment (amdrexu) [inline llpc/patch/llpcPatchInOutImportExport.cpp]: Better to rename it "$deviceIdx". "dev" has more meanings.

File: llpc/util/llpcElfWriter.cpp [hunk 4]
```diff
@@ -1283,6 +1305,10 @@ void FixUpRelocations(
         {
             FixUpDescriptorOffsetRelocation(pWriter, pContext, reloc, isGraphicsPipeline);
         }
+        else if (strcmp(reloc.name, "$devIdx") == 0)
```
Comment (amdrexu) [inline llpc/util/llpcElfWriter.cpp]: The same as above: devIdx -> deviceIdx

**Comment by JaxLinAMD** [comment]
there are conlicts need to resolve

**Comment by amdvlk-admin** [comment]
### Test summary for commit 978abc677343ff990376b6d76bdd86ed2c4cabde
<details open><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master b601c88aeca7a7b08becb3d32709de383c8ee428</li>
<li>METROHASH: amd-master 2b6fee002db6cc92345b02aeee963ebaaf4c0e2f</li>
<li>PAL: dev bb7398b1c7509a30ba4f7d947b5ad5549a4d58db</li>
<li>SPVGEN: dev dc2b146de03ef54f188172d8fa2f9be708cbfaca</li>
<li>XGL: dev ffd39443fec0ce821b7702173eba7184855c7b5e</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev 84d753b2c2274922020557981a17d1667378f746</li>
</ul></details>

<details open><summary>ShaderDB test</summary><ul>

<li>Passed: 958/958 (100.0%)</li>
<li>Failed: 0/958 (0.0%)</li>
</ul></details>

<details open><summary>CTS tests (built with branch Vulkan-CTS-1.2.0.2-fetch-sources-fix)</summary><ul>


<details open><summary>Ubuntu 18.04, Ellesmere</summary><ul>

<li>Passed: 103136/103137 (100.0%)</li>
<li><details><summary>Failed: 1/103137 (0.0%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.pipeline.push_constant.graphics_pipeline.count_2_shaders_vert_frag
Stack trace: Image mismatch
```
</p></details></li>


</ul></details>

<details open><summary>Ubuntu 18.04, Navi10</summary><ul>

<li>Passed: 27239/27240 (100.0%)</li>
<li><details><summary>Failed: 1/27240 (0.0%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.pipeline.push_constant.graphics_pipeline.count_2_shaders_vert_frag
Stack trace: Timeout or crahsed
```
</p></details></li>


</ul></details>

<details open><summary>Rhel 7.6, Vega10</summary><ul>

<li>Passed: 27239/27240 (100.0%)</li>
<li><details><summary>Failed: 1/27240 (0.0%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.pipeline.push_constant.graphics_pipeline.count_2_shaders_vert_frag
Stack trace: Timeout or crahsed
```
</p></details></li>


</ul></details>


</ul></details>

**Comment by JacobHeAMD** [comment]
It seems this change causes GPU hang. The agents have been rebooted, please check the change first before trigger the test again.

**Comment by amdvlk-admin** [comment]
### Test summary for commit 21a6a5512a4c5f5b916d69c0dc64c112ff580f03
<details open><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master b601c88aeca7a7b08becb3d32709de383c8ee428</li>
<li>METROHASH: amd-master 2b6fee002db6cc92345b02aeee963ebaaf4c0e2f</li>
<li>PAL: dev b5e6dde2a0aa5dd2ce6d794e58d89ef715570f2d</li>
<li>SPVGEN: dev e843618d71c8263030551e7692b7683f20f627fe</li>
<li>XGL: dev 2c44bbc7b58efba12c501f71f5b268f83c1bdad5</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev 5ae05ed8bd4f232c8580db96c021cad1e5c300e3</li>
</ul></details>

<details open><summary>ShaderDB test</summary><ul>

<li>Passed: 958/958 (100.0%)</li>
<li>Failed: 0/958 (0.0%)</li>
</ul></details>

<details open><summary>CTS tests (built with branch Vulkan-CTS-1.2.0.2-fetch-sources-fix)</summary><ul>


<details open><summary>Ubuntu 18.04, Ellesmere</summary><ul>

<li>Passed: 103142/103142 (100.0%)</li>
<li>Failed: 0/103142 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Navi10</summary><ul>

<li>Passed: 103198/103198 (100.0%)</li>
<li>Failed: 0/103198 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 7.6, Vega10</summary><ul>

<li>Passed: 103198/103198 (100.0%)</li>
<li>Failed: 0/103198 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by JaxLinAMD** [comment]
@csyonghe please resolve conflicts

**Comment by amdvlk-admin** [comment]
### Test summary for commit b522ae35bf8bb64eeb95d253a3a59ba77f06e5c2
<details open><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master b601c88aeca7a7b08becb3d32709de383c8ee428</li>
<li>METROHASH: amd-master 2b6fee002db6cc92345b02aeee963ebaaf4c0e2f</li>
<li>PAL: dev b5e6dde2a0aa5dd2ce6d794e58d89ef715570f2d</li>
<li>SPVGEN: dev e843618d71c8263030551e7692b7683f20f627fe</li>
<li>XGL: dev 2c44bbc7b58efba12c501f71f5b268f83c1bdad5</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev f02738a5cf6c158c8c5b25a5048951f3059cabe2</li>
</ul></details>

<details open><summary>ShaderDB test</summary><ul>

<li>Passed: 958/958 (100.0%)</li>
<li>Failed: 0/958 (0.0%)</li>
</ul></details>

<details open><summary>CTS tests (built with branch Vulkan-CTS-1.2.0.2-fetch-sources-fix)</summary><ul>


<details open><summary>Ubuntu 18.04, Ellesmere</summary><ul>

<li>Passed: 103142/103142 (100.0%)</li>
<li>Failed: 0/103142 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Navi10</summary><ul>

<li>Passed: 103198/103198 (100.0%)</li>
<li>Failed: 0/103198 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 7.6, Vega10</summary><ul>

<li>Passed: 103198/103198 (100.0%)</li>
<li>Failed: 0/103198 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by JaxLinAMD** [comment]
merge process is started, please don't force push.

**Comment by JaxLinAMD** [comment]
@csyonghe  please resolve conflicts.

**Comment by amdvlk-admin** [comment]
### Test summary for commit 3b8afc9edb500d8f1d54cdbb506711ca0c064749
<details open><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master b601c88aeca7a7b08becb3d32709de383c8ee428</li>
<li>METROHASH: amd-master 2b6fee002db6cc92345b02aeee963ebaaf4c0e2f</li>
<li>PAL: dev e2f8eb0ffe21751b300995e795965d37e8d6dd67</li>
<li>SPVGEN: dev e843618d71c8263030551e7692b7683f20f627fe</li>
<li>XGL: dev 2c44bbc7b58efba12c501f71f5b268f83c1bdad5</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev f02738a5cf6c158c8c5b25a5048951f3059cabe2</li>
</ul></details>

<details open><summary>ShaderDB test</summary><ul>

<li>Passed: 958/958 (100.0%)</li>
<li>Failed: 0/958 (0.0%)</li>
</ul></details>

<details open><summary>CTS tests (built with branch Vulkan-CTS-1.2.0.2-fetch-sources-fix)</summary><ul>


<details open><summary>Ubuntu 18.04, Ellesmere</summary><ul>

<li>Passed: 103142/103142 (100.0%)</li>
<li>Failed: 0/103142 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Navi10</summary><ul>

<li>Passed: 103198/103198 (100.0%)</li>
<li>Failed: 0/103198 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 7.6, Vega10</summary><ul>

<li>Passed: 103198/103198 (100.0%)</li>
<li>Failed: 0/103198 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by JaxLinAMD** [comment]
merge process is started, please don't force push :)

**Comment by csyonghe** [comment]
Should I resolve the conflict here?

**Comment by JaxLinAMD** [comment]
@csyonghe yes, please resolve conflicts here.

**Pull request merged**